# Vector Database (Bancos de Dados vetoriais) e Embeddings Models

## Introdução Vector Database : 

Podemos definir Banco de Dados Veorial como uma solução de gerenciamento de dados projetada para otimizar a recuperação e o armazenamento de informações em formato de vetores.
Essa abordagem utiliza algoritmos específicos para transformar os dados em um conjunto de números (vetores), que são representados em um espaço vetorial.
Essa solução é muito importante para aplicações de Machine Learning e Inteligência Artificial, pois permite encontrar itens semelhantes com base na proximidade vetorial, viabilizando funcionalidades como busca semântica, sistemas de recomendação e análise de padrões de forma eficiente e escalável.

## Weaviate:

Existem diversos tipo de soluções de banco de dados vetoriais disponiveis no mercado, mas para nossos testes, utilizaremos o Weaviate(https://weaviate.io/), pois, além de ser open-source, ele oferece excelente suporte para buscas semânticas e dados estruturados, além de integração com modelos de embeddings populares.

## O que são Modelos de Embeddings?

Bascimente os Embeddings são representações vetoriais de dados que capturam seus significados ou características de forma numérica em um espaço de alta dimensão. 

É um conceito amplamente utilizados em Machine Learning, especialmente em tarefas de processamento de linguagem natural (PLN) e visão computacional.

### **Objetivos:**

Como dito anteriormente, o principal objetivo dos modelos de embeddings é representar dados textuais, visuais ou numéricos em um formato numérico, preservando suas relações semânticas. Em outras palavras, isso significa que a representação vetorial das palavras "cachorro" e "gato" será semanticamente próxima, enquanto "cachorro" e "carro" estarão mais distantes no espaço vetorial.

### Exemplos de Modelos de Embeddings:

**Modelos populares de embeddings:**
- **Word2Vec:** Gera embeddings estáticos para palavras com base em coocorrência no texto.
- **GloVe:** Similar ao Word2Vec, mas utiliza estatísticas globais de um corpus.
- **BERT:** Modelo contextualizado que gera embeddings específicos para cada palavra em uma sentença, considerando o contexto.
- **CLIP:** Utilizado para embeddings de texto e imagens, projetados no mesmo espaço vetorial.
- **Sentence Transformers (SBERT):** Ideal para embeddings de frases e documentos.

Cada modelo é projetado para atender a diferentes tipos de tarefas. Escolha o mais adequado com base no problema a ser resolvido.

### Geração de Embeddings na Prática

Nesta sessão, vamos explorar modelos de embeddings e realizar análises para entender melhor o conceito de análise semântica e contextual entre diferentes tipos de dados.

In [59]:
from sentence_transformers import SentenceTransformer, util

In [7]:
# Carregando o Modelo
model = SentenceTransformer('all-MiniLM-L6-v2')

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

C:\Users\dtiDigital\Documents\Codice Juridico\.venv\Lib\site-packages\huggingface_hub\file_download.py:140: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\dtiDigital\.cache\huggingface\hub\models--sentence-transformers--all-MiniLM-L6-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [89]:
# Definição das frases que usaremos como exemplo
frases = [
    "O cachorro está no parque.",
    "O gato subiu na árvore.",
    "Carros são meios de transporte."
]

In [90]:
# Gerando embeddings
embeddings = model.encode(frases)

In [91]:
# Mostrando resultados
for i, frase in enumerate(frases):
    print(f"Frase: {frase}")
    print(f"Embedding: {embeddings[i][:5]}... (dimensão: {len(embeddings[i])})\n")

Frase: O cachorro está no parque.
Embedding: [-0.03486934  0.00535641 -0.02055818 -0.01926387 -0.08245736]... (dimensão: 384)

Frase: O gato subiu na árvore.
Embedding: [-0.09742771  0.05417777 -0.06127026  0.02175123 -0.07582309]... (dimensão: 384)

Frase: Carros são meios de transporte.
Embedding: [ 0.01851568  0.03419538 -0.00255625  0.05014872 -0.0176652 ]... (dimensão: 384)



Parece aleatório, né? Vamos entender melhor o que esses números "aleatórios" representam.
Acredito que o ponto mais importe para entender é que o vetor deve ser analisado como um todo, o que quero dizer que apensar do conter 5 números e as frases conterem 5 palavras, esses valores não representam diretamente o significado de palavras específicas da frase. Em outras palavras, não podemos olhar para um único número no vetor e dizer que ele corresponde a uma palavra ou ideia.

Portanto, não podemos interpretar diretamente que palavras como "gato" e "cachorro" são semelhantes só porque um valor específico nos embeddings é próximo. O vetor completo (todos as 384 dimensões) deve ser considerado para avaliar semelhança.

O interessante de entendermos esse conceito é que começamos a entender como os famosos LLMs funcionam, mas o que isso quer dizer? Pense no seguinte exemplo: em português, a palavra "cachorro" pode ter vários significados. Ela pode se referir ao animal, mas também pode ser usada de forma figurada para descrever alguém bravo, alguém que não presta, ou até em expressões regionais. Agora, imagine que o computador não raciocina e não entende palavras ou significados como nós, humanos. Para ele, tudo se resume a números. Então, para que ele consiga distinguir entre o "cachorro" animal e o "cachorro" de uma expressão figurada, precisamos fornecer um modelo matemático que capture essas nuances. É aqui que os embeddings entram: eles traduzem as frases em vetores numéricos que representam o contexto e o significado subjacente.

Abaixo, realizaremos o embedding de algumas frases para compreender melhor as diferenças contextuais e semânticas.

In [96]:
# Definição das frases que usaremos como exemplo
frases = [
    "O cachorro está no quintal.", 
    "Ele é um cachorro bravo, ninguém consegue conversar com ele.",  
    "Aquele homem é um cachorro, traiu a namorada de novo.", 
    "O cachorro está no quintal brincando com os brinquedos dele."
]

In [97]:
# Gerando embeddings
embeddings = model.encode(frases)

In [98]:
# Calculando similaridades
similaridade_1_2 = util.cos_sim(embeddings[0], embeddings[1])
similaridade_1_3 = util.cos_sim(embeddings[0], embeddings[2])
similaridade_2_3 = util.cos_sim(embeddings[1], embeddings[2])
similaridade_1_4 = util.cos_sim(embeddings[0], embeddings[3])

# Mostrando resultados
print(f"Similaridade entre frases 1 e 2: {similaridade_1_2.item():.4f}")
print(f"Similaridade entre frases 1 e 3: {similaridade_1_3.item():.4f}")
print(f"Similaridade entre frases 2 e 3: {similaridade_2_3.item():.4f}")
print(f"Similaridade entre frases 1 e 4: {similaridade_1_4.item():.4f}")

Similaridade entre frases 1 e 2: 0.5939
Similaridade entre frases 1 e 3: 0.5420
Similaridade entre frases 2 e 3: 0.5886
Similaridade entre frases 1 e 4: 0.7778


Depois de executar essa celulas do notebook acredito que fique mais facil entender um pouco sobre o que comentei a cima, mas antes é importante resultar o significado de 2 palavras e suas diferenças: 

- **Semantica:** Semântica é o estudo do significado das palavras, frases e sentenças em um idioma. Refere-se à relação entre os símbolos linguísticos (como palavras e frases) e os conceitos ou significados que eles representam.
- **Contexto:** contexto refere-se ao conjunto de informações ou circunstâncias que rodeiam um evento, situação, ou palavra, e que ajudam a interpretar o seu significado de forma adequada. No caso de linguagens, o contexto pode incluir palavras adjacentes, o tópico da conversa ou mesmo a intenção do autor. Exemplo: A palavra "banco" pode significar uma instituição financeira ou um objeto para sentar, dependendo do contexto.